ros2-library/src/main/java/us/ihmc/ros2/ROS2NodeBuilder.java

```diff
@@ -0,0 +1,260 @@
+package us.ihmc.ros2;
+
+import us.ihmc.log.LogTools;
+import us.ihmc.pubsub.attributes.ParticipantProfile;
+import us.ihmc.util.PeriodicNonRealtimeThreadSchedulerFactory;
+import us.ihmc.util.PeriodicThreadSchedulerFactory;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.util.Properties;
+
+public class ROS2NodeBuilder
+{
+   public enum TransportSpecial
+   {
+      SHARED_MEMORY_ONLY, LOOPBACK_ADDRESS_ONLY, UDPV4_ONLY
+   }
+
+   private static final int UNSET_DOMAIN_ID = -1;
+
+   private int domainId = UNSET_DOMAIN_ID;
+   private String namespace = "/us/ihmc";
+   private boolean useSharedMemory = true;
+   private InetAddress[] addressRestriction;
+
+   private boolean parseProperties = true;
+   private boolean parseEnvironment = true;
+   private boolean parseNetworkParametersConfig = true;
+
+   @Nullable
+   private TransportSpecial transportSpecial;
+
+   public ROS2NodeBuilder domainId(int domainId)
+   {
+      this.domainId = domainId;
+      return this;
+   }
+
+   public ROS2NodeBuilder namespace(String namespace)
+   {
+      this.namespace = namespace;
+      return this;
+   }
+
+   public ROS2NodeBuilder useSharedMemory(boolean useSharedMemory)
+   {
+      this.useSharedMemory = useSharedMemory;
+      return this;
+   }
+
+   public ROS2NodeBuilder addressRestriction(InetAddress... addressRestriction)
+   {
+      this.addressRestriction = addressRestriction;
+      return this;
+   }
+
+   public ROS2NodeBuilder parseProperties(boolean parseProperties)
+   {
+      this.parseProperties = parseProperties;
+      return this;
+   }
+
+   public ROS2NodeBuilder parseEnvironment(boolean parseEnvironment)
+   {
+      this.parseEnvironment = parseEnvironment;
+      return this;
+   }
+
+   public ROS2NodeBuilder parseNetworkParametersConfig(boolean parseNetworkParametersConfig)
+   {
+      this.parseNetworkParametersConfig = parseNetworkParametersConfig;
+      return this;
+   }
+
+   public ROS2NodeBuilder transportSpecial(@Nullable TransportSpecial transportSpecial)
+   {
+      this.transportSpecial = transportSpecial;
+      return this;
+   }
+
+   private ParticipantProfile buildProfile()
+   {
+      ParticipantProfile profile = ParticipantProfile.create();
+
+      // Set up ROS Domain ID
+      {
+         if (domainIDValid(domainId))
+         {
+            LogTools.info("Using a programmatically set ROS Domain ID: {}", domainId);
+         }
+         else
+         {
+            // Try to find a ROS Domain ID
+            domainId = findDomainID(parseNetworkParametersConfig, parseProperties, parseEnvironment);
```

Comment:
What if we do this whole block in the constructor of ROS2Node, only if the domain ID is not set yet when it's passed in? Then only one print per node creation for domain ID decisions.